### PR TITLE
Add failing test for saving with a new sort_order

### PR DESCRIPTION
**Pull request by Ian-Foote**
Fix `IntegrityError` in `Orderable.save` when `qs.update` violates a `unique` constraint.

I used [hypothesis](https://hypothesis.readthedocs.org/en/latest/) because `Orderable.save` is very complex and has potentially many breaking edge-cases hidden away.

### CODE REVIEW

File: orderable/tests/test_models.py [hunk 1]
```diff
@@ -207,3 +209,10 @@ def test_changing_parent(self):
         subtask_2.task = task
         subtask_2.save()
         self.assertSequenceEqual(task.subtask_set.all(), [subtask, subtask_2])
+
+    @given(lists(integers(min_value=1), min_size=1, unique=True))
+    def test_save_subtask(self, sort_orders):
+        task = Task.objects.create()
+
+        for order in sort_orders:
+            SubTask.objects.create(task=task, sort_order=order)
```
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: The minimal failing examples all look like this:

```python
sort_orders = [n, n + 1, 1]
```

The `create` fails in `Orderable._save` when it tries to run `.update(sort_order=F('sort_order') + 1)` on the first two `SubTask` instances.

I think the problem is that the first `SubTask` with `sort_order == n` is incremented to `sort_order == n + 1` which conflicts with the second `SubTask` because of the `unique_together` between `sort_order` and `task`.

Comment (Ian-Foote) [inline orderable/tests/test_models.py]: This doesn't occur with the `Task` model because it doesn't have any `unique` constraints.

**Comment by Ian-Foote** [comment]
@meshy I'm not sure how best to fix this. Any ideas?

File: orderable/models.py [hunk 2]
```diff
@@ -67,6 +67,20 @@ def prev(self):
...
+        Increment the sort_order in a queryset.
+
+        Handle IntegrityErrors caused by unique constraints.
+        """
+        try:
+            with transaction.atomic():
+                qs.update(sort_order=models.F('sort_order') + 1)
+        except IntegrityError:
+            for obj in qs.order_by('-sort_order'):
+                qs.filter(pk=obj.pk).update(sort_order=models.F('sort_order') + 1)
```
Comment (Ian-Foote) [inline orderable/models.py]: This is ugly, but it works...

Unfortunately, Django doesn't have support for [`DEFERRABLE` unique constraints](http://www.postgresql.org/docs/current/static/sql-set-constraints.html). Ideally, we'd be able to set the constraint as `UNIQUE DEFERRABLE INITIALLY IMMEDIATE` and then run the `update` with `DEFERRED` set.

Comment (meshy) [inline orderable/models.py]: This doesn't seem right to me -- why is there an integrity error in the first place? The "shift" should have cleared the way, no?

Comment (meshy) [inline orderable/models.py]: (Please bear in mind that I don't really understand the failing case here.)

Comment (Ian-Foote) [inline orderable/models.py]: Say we have a `qs` where `sort_order` looks like this: `1, 2, 3, 4`.
We add a new item with `sort_order == 2`.
`Orderable._save` creates `to_shift` which looks like: `2, 3, 4`.
The old code then tried to call `to_shift.update(sort_order = F('sort_order') + 1)`.
This works fine without any `unique` constraints, but if `sort_order` is `unique_together` with another field (as in `SubTask`) the `update` will raise `IntegrityError` when it increments `2` to `3`. The internal state of `to_shift` is now `3, 3, 4`, which is invalid.

If Django (and the db) supported `DEFERRED`, we could defer the unique check until the `update` had completed (`3, 4, 5`) and everything would work as in the case without `unique`.

Comment (meshy) [inline orderable/models.py]: Sounds like a manifestation of https://github.com/incuna/django-orderable/issues/7

Comment (Ian-Foote) [inline orderable/models.py]: It definitely looks related to #26 / #33.

Comment (Ian-Foote) [inline orderable/models.py]: Anyway, I agree this isn't really the "right" fix, but I don't think I can implement the "right" fix without `DEFERRED`.

**Comment by Ian-Foote** [comment]
@KevinEtienne @adam-thomas Review?

File: orderable/models.py (hunk 2 above)
Comment (KevinEtienne) [inline orderable/models.py]: What about defining the new sort order by: `Max(sort_order) + sort_order + 1`? Would that work?

File: orderable/tests/test_models.py [hunk 3]
```diff
@@ -207,3 +213,13 @@ def test_changing_parent(self):
         subtask_2.task = task
         subtask_2.save()
         self.assertSequenceEqual(task.subtask_set.all(), [subtask, subtask_2])
+
+    @given(lists(integers(min_value=1), min_size=1, unique=True))
```
Comment (adam-thomas) [inline orderable/tests/test_models.py]: What does this do?

File: orderable/tests/test_models.py [hunk 4]
```diff
@@ -207,3 +213,13 @@ def test_changing_parent(self):
...
+    def test_save_subtask(self, sort_orders):
+        task = Task.objects.create()
+
+        for order in sort_orders:
+            subtask = SubTask.objects.create(task=task, sort_order=order)
+
+        subtask.sort_order = 2
+        subtask.save()
```
Comment (adam-thomas) [inline orderable/tests/test_models.py]: Does this test fail or not?  Slightly confused by the issue title.  Travis seems to think it's all good.

File: orderable/tests/test_models.py (hunk 3 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: Populates the `sort_order` parameter with a list of positive integers.

File: setup.cfg [hunk 5]
```diff
@@ -2,7 +2,7 @@
 universal = 1
 [flake8]
 application-import-names = orderable
-ignore = D100,D101,D102,D203,D204
+ignore = D100,D101,D102,D104,D105,D203,D204
```
Comment (adam-thomas) [inline setup.cfg]: good old `flake8-docstrings`.  What do you (and the rest of @incuna/backend) think about just not using it?  We're ignoring quite a lot of the warnings here - is what's left even all that useful?

File: orderable/tests/test_models.py (hunk 4 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: `Orderable.save` used to raise `IntegrityError` in some cases. This PR fixes that.

File: orderable/tests/test_models.py (hunk 3 above)
Comment (adam-thomas) [inline orderable/tests/test_models.py]: Rather than installing Hypothesis, wouldn't it be better just to do `sort orders = range(1, n)`?

File: setup.cfg (hunk 5 above)
Comment (Ian-Foote) [inline setup.cfg]: I'm happy to remove it, but that might as well be a different PR.

File: orderable/tests/test_models.py (hunk 3 above)
Comment (adam-thomas) [inline orderable/tests/test_models.py]: (With `sort_orders.shuffle()` if you need them to be randomly distributed.  Or writing out a short list by hand.)

File: setup.cfg (hunk 5 above)
Comment (Ian-Foote) [inline setup.cfg]: Though actually, there's still a lot of errors we do catch: http://pep257.readthedocs.org/en/latest/error_codes.html

Comment (adam-thomas) [inline setup.cfg]: Yeah, no rush.  (We'll want to remove it in our other projects as well, if everyone's on board with the idea.)

File: orderable/tests/test_models.py (hunk 4 above)
Comment (adam-thomas) [inline orderable/tests/test_models.py]: Great, ta.

File: setup.cfg (hunk 5 above)
Comment (Ian-Foote) [inline setup.cfg]: It's mostly missing docstrings we ignore, along with whether there are blank lines before/after class docstrings.

File: orderable/tests/test_models.py (hunk 3 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: This is much more thorough. I don't trust us to catch all the edge cases without hypothesis.

File: setup.cfg (hunk 5 above)
Comment (adam-thomas) [inline setup.cfg]: Looking at the list of errors, I think there are only one or two that are particularly significant.  A lot of them are things we all do by habit, although they're still useful for newcomers in the future so that's not a good reason to remove it.  (There's also a couple of stupid ones like "First line should be in imperative mood" :P)

File: orderable/tests/test_models.py (hunk 3 above)
Comment (adam-thomas) [inline orderable/tests/test_models.py]: Thorough how?  Does it produce a different list each time you run the test?  That seems really bad.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: Also, how do you know you can trust Hypothesis in the first place?  Can we see the data it generates?

File: orderable/models.py (hunk 2 above)
Comment (Ian-Foote) [inline orderable/models.py]: Nice idea. I tried a few variants of it - the first just led to `IntegrityError` for a slightly more complex example. The second caused `sort_order` to increase exponentially - I get `django.db.utils.DataError` eventually.

File: setup.cfg (hunk 5 above)
Comment (Ian-Foote) [inline setup.cfg]: We do do most of these by habit, but that's the point of linting - enforcing good habits.

File: orderable/tests/test_models.py (hunk 3 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: https://hypothesis.readthedocs.org/en/latest/

Comment (meshy) [inline orderable/tests/test_models.py]: It can't hurt to add the `@example` of the text you're expecting to fail here.

File: orderable/tests/test_models.py (hunk 4 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: I added a docstring.

File: orderable/models.py (hunk 2 above)
Comment (KevinEtienne) [inline orderable/models.py]: cool thanks :)

File: orderable/tests/test_models.py (hunk 3 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: Good idea.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: I've read the Hypothesis docs before, last time we were talking about using it.  It looked a bit rubbish to me at the time.  I'd appreciate an explanation of why we're suddenly using it and what its benefits are, please.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: My main concern here is that by hiding test data behind a wall of magic, it's going to be more difficult to debug and edit tests, and I don't trust Hypothesis to follow code updates perfectly or cover the space of data it promises to cover.  Especially when it comes to more complex objects.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: "I strongly recommend not using `TransactionTestCase` unless you really have to. Because Hypothesis runs this in a loop the performance problems it normally has are significantly exacerbated and your tests will be really slow."  This impacts a majority of our tests, including anything based on a `RequestTestCase`, for instance.

Comment (Ian-Foote) [inline orderable/tests/test_models.py]: I'm not proposing using it everywhere. I think it is appropriate here because `Orderable.save` is very complex with a ton of possible breaking edge-cases.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: "Hypothesis creates saved models. This will run inside your testing transaction when using the test runner, but if you use the dev console this will leave debris in your database."  That's potentially going to break tests that use queryset retrieval methods and assert the results.

Looking at these docs doesn't exactly fill me with confidence.  There are a lot of ifs, buts, caveats, and limitations.  I seem to recall this being among the reasons we didn't start using it when it first came to our attention - the project just doesn't seem ready or mature enough, at least in its Django support.

File: orderable/tests/test_models.py (hunk 4 above)
Comment (adam-thomas) [inline orderable/tests/test_models.py]: I mean the issue description, it's misleading (but good idea, thanks!)

File: orderable/tests/test_models.py (hunk 3 above)
Comment (Ian-Foote) [inline orderable/tests/test_models.py]: That's just a warning about using `hypothesis` in `python manage.py shell` I think. The standard usage in tests is well supported.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: > I'm not proposing using it everywhere. I think it is appropriate here because Orderable.save is very complex with a ton of possible breaking edge-cases.

OK.  Do you know what those edge cases might look like?  Is there a way to use Hypothesis to help find them if they exist, and then write explicit tests to catch them?

In your test here, Hypothesis generates a set of randomish numbers to use as `sort_order` values, but the real error/problem is the `unique_together` with `SubTask` that you described earlier.  If the edge cases are anything like that, I don't think Hypothesis is going to help us - we can't use it to generate random database schemas or anything like that.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: > That's just a warning about using hypothesis in python manage.py shell I think. The standard usage in tests is well supported.

That's good at least.  There are unfortunately lots more caveats though.  (I was looking at https://hypothesis.readthedocs.org/en/latest/django.html)

Comment (Ian-Foote) [inline orderable/tests/test_models.py]: It found the bug(s) this fixes.

Comment (Ian-Foote) [inline orderable/tests/test_models.py]: Ok, I see your point that I did need to test a model that has the `unique` constraint. But that's something `django-orderable` claims to explicitly support.

Comment (Ian-Foote) [inline orderable/tests/test_models.py]: I'll also note that it found an overflow in the `sort_order` field with one of my attempts to improve `_update`. This is the sort of edge-case that I would not have found with a simple test.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: Found the bugs how?  It looks like it improved on the hardcoded list in the test that was there already, but wouldn't changing that to `[1,2,3,4]` as you described earlier have achieved the same thing?  Did you get that from the test output?  Do you think you would have been able to find this error using Hypothesis without that test already in place (or without writing it yourself)?

And yes - the point being that there probably aren't many more edge cases or failures in the exact list of `sort_orders` you have (there are only so many different relevant situations you can construct), but there might be plenty of edge cases or failures elsewhere that this test will never hit.  Adding Hypothesis here doesn't guarantee much.

Comment (adam-thomas) [inline orderable/tests/test_models.py]: > I'll also note that it found an overflow in the sort_order field with one of my attempts to improve _update. This is the sort of edge-case that I would not have found with a simple test.

That sounds good, although I'm a bit surprised/alarmed that `sort_order` can overflow at all outside of a case where we've basically filled up the database, at which point we have other issues.  Theoretically any field with a limited size can overflow in that situation, so I don't know if there's any value in designing around it.

Is there a test for this?  What else did it find?  If I'm missing information about what you've actually done here, please post it either in the issue or in the code itself.

Comment (Ian-Foote) [inline orderable/tests/test_models.py]: https://github.com/incuna/django-orderable/pull/36#discussion_r49444290

**Comment by Ian-Foote** [comment]
@adam-thomas @KevinEtienne Merge?

**Comment by KevinEtienne** [comment]
Merging, other issues if needed can be done in other PRs

**Comment by KevinEtienne** [comment]
`changelog.md` :p
